queue.go

```diff
@@ -57,50 +55,47 @@ type Consumer interface {
 	Close() error
 }
 
-// Producer wraps the producer implementation details
+// Producer wraps the producer implementation details. Producer implementations
+// must support sync and async production.
 type Producer interface {
-	model.BatchProcessor
+	// Produce produces N records. If the Producer is synchronous, waits until
+	// all records are produced, otherwise, returns as soon as the records are
+	// stored in the producer buffer, or when the records are produced to the
+	// queue if sync producing is configured.
+	// If the context has been enriched with metadata, each entry will be added
+	// as a record's header.
+	// Produce takes ownership of Record and any modifications after Produce is
+	// called may cause an unhandled exception.
+	Produce(ctx context.Context, rs ...Record) error
 	// Healthy returns an error if the producer isn't healthy.
 	Healthy(ctx context.Context) error
 	// Close closes the producer.
 	Close() error
 }
 
-// Topic represents a destination topic where to produce a message/record.
-type Topic string
-
-// TopicRouter is used to determine the destination topic for an model.APMEvent.
-type TopicRouter func(event model.APMEvent) Topic
+// Record wraps a record's value with the topic where it's produced / consumed.
+type Record struct {
```

Comment:
Instead of enriching the context with a map via `queuecontext.WithMetadata(ctx, map[string]string{}`, each record could have its own header, then it's up to the client to use the same headers for all the messages or have individual headers for each message.